[QUESTION] Why mcp messages are logged as error using stdio transport

Question

This is a question and potentially an issue to be fixed. Why are the messages of a MCP process launched logged as ERROR by the `ProcessStderrHandler`  using stdio transport?

```
2026-04-08 07:45:28,310 DEBUG [dev.lan.mcp.cli.tra.std.ProcessStderrHandler] (executor-thread-2) [ERROR] WARNING: sun.misc.Unsafe::objectFieldOffset will be removed in a future release
2026-04-08 07:45:28,502 DEBUG [dev.lan.mcp.cli.tra.std.ProcessStderrHandler] (executor-thread-2) [ERROR] __  ____  __  _____   ___  __ ____  ______ 
2026-04-08 07:45:28,503 DEBUG [dev.lan.mcp.cli.tra.std.ProcessStderrHandler] (executor-thread-2) [ERROR]  --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
2026-04-08 07:45:28,503 DEBUG [dev.lan.mcp.cli.tra.std.ProcessStderrHandler] (executor-thread-2) [ERROR]  -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
2026-04-08 07:45:28,503 DEBUG [dev.lan.mcp.cli.tra.std.ProcessStderrHandler] (executor-thread-2) [ERROR] --\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2026-04-08 07:45:28,503 DEBUG [dev.lan.mcp.cli.tra.std.ProcessStderrHandler] (executor-thread-2) [ERROR] 2026-04-08 07:45:28,500 INFO  [io.qua.mcp.ser.fil.MCPServerFS] (main) Starting file server with paths: [/Users/cmoullia/code/_temp/filesystem-assistant/playground]
2026-04-08 07:45:28,505 DEBUG [dev.lan.mcp.cli.tra.std.ProcessStderrHandler] (executor-thread-2) [ERROR] 2026-04-08 07:45:28,505 INFO  [io.quarkus] (main) mcp-server-filesystem 1.0.0.CR4 on JVM (powered by Quarkus 3.20.1) started in 0.419s. Listening on: 
2026-04-08 07:45:28,505 DEBUG [dev.lan.mcp.cli.tra.std.ProcessStderrHandler] (executor-thread-2) [ERROR] 2026-04-08 07:45:28,505 INFO  [io.quarkus] (main) Profile prod activated. 
2026-04-08 07:45:28,506 DEBUG [dev.lan.mcp.cli.tra.std.ProcessStderrHandler] (executor-thread-2) [ERROR] 2026-04-08 07:45:28,505 INFO  [io.quarkus] (main) Installed features: [cdi, mcp-server-sse, mcp-server-stdio, qute, smallrye-context-propagation, vertx]
...
```

> Also to add, this is a thing of the `quarkus-mcp-server` extension, langchain4j doesn't have anything to do with this, it just forwards those stderr logs into its own logging infrastructure.

Perhaps but I got such `[ERROR] text messages` when we set the logging level to `DEBUG` for langchain4j code => `quarkus.log.category."dev.langchain4j".level=DEBUG`

> They are routed on STDERR.
> Otherwise, they would interfere with the STDIO protocol used by MCP, because logging messages would be on STDOUT, and the MCP server would be confused by those log messages unrelated to its work.

Many thanks for the clarification. I would like to suggest nevertheless to change the text logged here as this is not true to log the starting messages as `[ERROR]`.

```java
// https://github.com/langchain4j/langchain4j/blob/74a5986c3b39fa875ae0ebc6a7433d32fb1a4204/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/stdio/ProcessStderrHandler.java#L26

    public void run() {
        try (InputStreamReader inputStreamReader = new InputStreamReader(process.getErrorStream())) {
            try (BufferedReader reader = new BufferedReader(inputStreamReader)) {
                String line;
                while ((line = reader.readLine()) != null) {
                    log.debug("[ERROR] {}", line);
                }
            } catch (IOException e) {
```

> Maybe changing the word `ERROR` to `STDERR` would help a bit?

I don't think that we should change the word. I would like to suggest instead when user set the log level to `DEBUG` that a message is logged before we start the MCP server using stdio mentioning how messages ae logged.

For the end users where they cannot change the log level and where the default is `INFO` such messages will not be displayed except if an error occurs but then the error should be certainly handled in a different way. See hereafter.

I did a test where I use the node `file-system` MCP server and where no `playground` folder exists containing a Hello.txt and then such a error is reported

```
2026-04-08 14:09:11,350 WARN  [dev.lan.mcp.McpToolProvider] (Quarkus Main Thread) Failed to retrieve tools from MCP server

Exception in CodeAssistantCommand.java:20
          18      @Override
          19      public void run() {                                                                                                                                                                       
        → 20          System.out.println(codeAssistantService.chat(task));                                                                                                                                  
          21      }                                                                                                                                                                                         
          22  }                                                                                                                                                                                             
                                                                                                                                                                                                            
                                                                                                                                                                                                            
                                                                                                                                                                                                            
Exception in CodeAssistantCommand.java:20                                                                                                                                                                   
          18      @Override                                                                                                                                                                                 
          19      public void run() {                                                                                                                                                                       
        → 20          System.out.println(codeAssistantService.chat(task));                                                                                                                                  
          21      }                                                                                                                                                                                         
          22  }                                                                                                                                                                                             
                                                                                                                                                                                                            
                                                                                                                                                                                                            
                                                                                                                                                                                                            
Exception in CodeAssistantCommand.java:20                                                                                                                                                                   
          18      @Override                                                                                                                                                                                 
          19      public void run() {                                                                                                                                                                       
        → 20          System.out.println(codeAssistantService.chat(task));                                                                                                                                  
          21      }                                                                                                                                                                                         
          22  }                                                                                                                                                                                             
                                                                                                                                                                                                            
: jakarta.enterprise.inject.CreationException: Error creating synthetic bean [kQV_Xi0MnUWxqNhOXO34PoeWdYQ]: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Operation cancelled due to transport failure: Process has exited                                                                                                                                      
        at dev.langchain4j.mcp.client.McpClient_kQV_Xi0MnUWxqNhOXO34PoeWdYQ_Synthetic_Bean.doCreate(Unknown Source)                                                                                         
        at dev.langchain4j.mcp.client.McpClient_kQV_Xi0MnUWxqNhOXO34PoeWdYQ_Synthetic_Bean.create(Unknown Source)                                                                                           
        at dev.langchain4j.mcp.client.McpClient_kQV_Xi0MnUWxqNhOXO34PoeWdYQ_Synthetic_Bean.create(Unknown Source)                                                                                           
        at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)                                                                                                   
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)                                                                                                                   
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)                                                                                                                   
        at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.c23(Unknown Source)                                                                               
        at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.computeIfAbsent(Unknown Source)                                                                   
        at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)                                                                                                                     
        at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:23)                                                                                                            
        at dev.langchain4j.mcp.client.McpClient_kQV_Xi0MnUWxqNhOXO34PoeWdYQ_Synthetic_ClientProxy.arc$delegate(Unknown Source)                                                                              
        at dev.langchain4j.mcp.client.McpClient_kQV_Xi0MnUWxqNhOXO34PoeWdYQ_Synthetic_ClientProxy.listTools(Unknown Source)                                                                                 
        at dev.langchain4j.mcp.McpToolProvider.provideTools(McpToolProvider.java:160)                                                                                                                       
        at io.quarkiverse.langchain4j.mcp.runtime.QuarkusMcpToolProvider.provideTools(QuarkusMcpToolProvider.java:42)                                                                                       
        at dev.langchain4j.service.tool.ToolProvider_sL_YcZTZI2rxkspvkYJxQT2XtaU_Synthetic_ClientProxy.provideTools(Unknown Source)                                                                         
        at io.quarkiverse.langchain4j.runtime.aiservice.AiServiceMethodImplementationSupport.doImplement0(AiServiceMethodImplementationSupport.java:264)                                                    
        at io.quarkiverse.langchain4j.runtime.aiservice.AiServiceMethodImplementationSupport.doImplement(AiServiceMethodImplementationSupport.java:205)                                                     
        at io.quarkiverse.langchain4j.runtime.aiservice.AiServiceMethodImplementationSupport.implement(AiServiceMethodImplementationSupport.java:180)                                                       
        at org.acme.ai.CodeAssistantService$$QuarkusImpl.chat(Unknown Source)                                                                                                                               
        at org.acme.ai.CodeAssistantService$$QuarkusImpl_ClientProxy.chat(Unknown Source)                                                                                                                   
        at org.acme.ai.CodeAssistantCommand.run(CodeAssistantCommand.java:20)                                                                                                                               
        at picocli.CommandLine.executeUserObject(CommandLine.java:2045)                                                                                                                                     
        at picocli.CommandLine.access$1500(CommandLine.java:148)                                                                                                                                            
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2469)                                                                                               
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2461)                                                                                                                                        
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2423)                                                                                                                                        
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)                                                                                                                    
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2425)                                                                                                                                       
        at io.quarkus.picocli.runtime.PicocliRunner$EventExecutionStrategy.execute(PicocliRunner.java:26)                                                                                                   
        at picocli.CommandLine.execute(CommandLine.java:2174)                                                                                                                                               
        at io.quarkus.picocli.runtime.PicocliRunner.run(PicocliRunner.java:40)                                                                                                                              
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:141)                                                                                                         
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:80)                                                                                                                                                  
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:51)                                                                                                                                                  
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)                                                                                                                                             
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)                                                                                            
        at java.base/java.lang.reflect.Method.invoke(Method.java:565)                                                                                                                                       
        at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:135)                                                                                                                  
        at java.base/java.lang.Thread.run(Thread.java:1474)                                                                                                                                                 
Caused by: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Operation cancelled due to transport failure: Process has exited                           
        at dev.langchain4j.mcp.client.DefaultMcpClient.initialize(DefaultMcpClient.java:169)                                                                                                                
        at dev.langchain4j.mcp.client.DefaultMcpClient.<init>(DefaultMcpClient.java:147)                                                                                                                    
        at dev.langchain4j.mcp.client.DefaultMcpClient$Builder.build(DefaultMcpClient.java:790)                                                                                                             
        at io.quarkiverse.langchain4j.mcp.runtime.McpRecorder$2.apply(McpRecorder.java:169)                                                                                                                 
        at io.quarkiverse.langchain4j.mcp.runtime.McpRecorder$2.apply(McpRecorder.java:72)                                                                                                                  
        at dev.langchain4j.mcp.client.McpClient_kQV_Xi0MnUWxqNhOXO34PoeWdYQ_Synthetic_Bean.createSynthetic(Unknown Source)                                                                                  
        ... 39 more                                                                                                                                                                                         
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Operation cancelled due to transport failure: Process has exited                                                       
        at java.base/java.util.concurrent.CompletableFuture.wrapInExecutionException(CompletableFuture.java:345)                                                                                            
        at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:440)                                                                                                           
        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2117)                                                                                                                
        at dev.langchain4j.mcp.client.DefaultMcpClient.initialize(DefaultMcpClient.java:166)                                                                                                                
        ... 44 more 
```

--- a/langchain4j-mcp-docker/src/main/java/dev/langchain4j/mcp/client/transport/docker/DockerResultCallback.java
+++ b/langchain4j-mcp-docker/src/main/java/dev/langchain4j/mcp/client/transport/docker/DockerResultCallback.java
@@ -1,18 +1,18 @@
 package dev.langchain4j.mcp.client.transport.docker;
 
+import static dev.langchain4j.internal.Utils.getOrDefault;
+
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.github.dockerjava.api.async.ResultCallback;
 import com.github.dockerjava.api.model.Frame;
 import com.github.dockerjava.api.model.StreamType;
 import dev.langchain4j.mcp.client.transport.McpOperationHandler;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import java.util.regex.Pattern;
-
-import static dev.langchain4j.internal.Utils.getOrDefault;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 class DockerResultCallback extends ResultCallback.Adapter<Frame> {
     private static final Logger LOG = LoggerFactory.getLogger(DockerResultCallback.class);
@@ -41,7 +41,7 @@ class DockerResultCallback extends ResultCallback.Adapter<Frame> {
     public void onNext(Frame frame) {
         String frameStr = new String(frame.getPayload());
         if (frame.getStreamType() == StreamType.STDERR) {
-            LOG.debug("[ERROR] {}", frameStr);
+            LOG.debug("[STDERR] {}", frameStr);
         } else if (frame.getStreamType() == StreamType.STDOUT) {
             this.send(frameStr);
         }
@@ -59,7 +59,7 @@ class DockerResultCallback extends ResultCallback.Adapter<Frame> {
     }
 
     private void send(String line) {
-        if (line !=null && !line.isBlank()) {
+        if (line != null && !line.isBlank()) {
             logAggregator.append(line);
 
             // we aggregate until we have a newline char at then end of the line

--- a/langchain4j-mcp-docker/src/test/java/dev/langchain4j/mcp/client/transport/docker/DockerResultCallbackTest.java
+++ b/langchain4j-mcp-docker/src/test/java/dev/langchain4j/mcp/client/transport/docker/DockerResultCallbackTest.java
@@ -1,0 +1,37 @@
+package dev.langchain4j.mcp.client.transport.docker;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+
+import com.github.dockerjava.api.model.Frame;
+import com.github.dockerjava.api.model.StreamType;
+import dev.langchain4j.mcp.client.transport.McpOperationHandler;
+import org.junit.jupiter.api.Test;
+import org.mockito.MockedStatic;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class DockerResultCallbackTest {
+
+    @Test
+    void shouldLabelDockerStderrFramesAsStderr() {
+        try (MockedStatic<LoggerFactory> loggerFactory = mockStatic(LoggerFactory.class)) {
+            Logger logger = mock(Logger.class);
+            loggerFactory
+                    .when(() -> LoggerFactory.getLogger(DockerResultCallback.class))
+                    .thenReturn(logger);
+            McpOperationHandler messageHandler = mock(McpOperationHandler.class);
+            DockerResultCallback callback = new DockerResultCallback(false, messageHandler);
+
+            callback.onNext(new Frame(StreamType.STDERR, "server log".getBytes(UTF_8)));
+
+            verify(logger).debug("[STDERR] {}", "server log");
+            verify(logger, never()).debug("[ERROR] {}", "server log");
+            verifyNoInteractions(messageHandler);
+        }
+    }
+}

--- a/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/stdio/ProcessStderrHandler.java
+++ b/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/stdio/ProcessStderrHandler.java
@@ -23,7 +23,7 @@ class ProcessStderrHandler implements Runnable, Closeable {
             try (BufferedReader reader = new BufferedReader(inputStreamReader)) {
                 String line;
                 while ((line = reader.readLine()) != null) {
-                    log.debug("[ERROR] {}", line);
+                    log.debug("[STDERR] {}", line);
                 }
             } catch (IOException e) {
                 // If this handler was closed, it means the MCP server process is shutting down,

--- a/langchain4j-mcp/src/test/java/dev/langchain4j/mcp/client/transport/stdio/ProcessStderrHandlerTest.java
+++ b/langchain4j-mcp/src/test/java/dev/langchain4j/mcp/client/transport/stdio/ProcessStderrHandlerTest.java
@@ -1,0 +1,36 @@
+package dev.langchain4j.mcp.client.transport.stdio;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.io.ByteArrayInputStream;
+import org.junit.jupiter.api.Test;
+import org.mockito.MockedStatic;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ProcessStderrHandlerTest {
+
+    @Test
+    void shouldLabelProcessStderrOutputAsStderr() {
+        try (MockedStatic<LoggerFactory> loggerFactory = mockStatic(LoggerFactory.class)) {
+            Logger logger = mock(Logger.class);
+            loggerFactory
+                    .when(() -> LoggerFactory.getLogger(ProcessStderrHandler.class))
+                    .thenReturn(logger);
+
+            Process process = mock(Process.class);
+            when(process.getErrorStream())
+                    .thenReturn(new ByteArrayInputStream(("server log" + System.lineSeparator()).getBytes()));
+            when(process.pid()).thenReturn(123L);
+
+            new ProcessStderrHandler(process).run();
+
+            verify(logger).debug("[STDERR] {}", "server log");
+            verify(logger, never()).debug("[ERROR] {}", "server log");
+        }
+    }
+}

